### PR TITLE
#1183:  Add Paged

### PR DESCRIPTION
**Pull request by andreoss**
Per #1183 
- Extract ctor as a separate class

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#1448](https://github.com/yegor256/cactoos/issues/1448) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), here is [why](https://www.0crat.com/footprint/C63314D6Z/44e5e2d4-9230-4aab-9cc9-12338d24e6ba); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @victornoel[/z](https://www.0crat.com/u/victornoel) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/org/cactoos/iterable/Paged.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                        return this.current.value().hasNext();
+                    }
+
+                    @Override
+                    public X next() {
+                        if (this.hasNext()) {
+                            return this.current.value().next();
+                        }
+                        throw new NoSuchElementException();
+                    }
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Paged.java:88]: @andreoss could you either create a `iterator.Paged` class that contains this code or add a todo to do so? This class would then be used here to implement `iterable.Paged`.

Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Paged.java:92]: @victornoel Todo added.

**Comment by victornoel** [comment]
@fabriciofx can you review?

File: src/main/java/org/cactoos/iterable/Paged.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    @SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
+    public <I extends Iterator<X>> Paged(
+        final Scalar<I> first, final Func<I, I> next
+    ) {
+        // @checkstyle AnonInnerLengthCheck (30 lines)
+        super(
+            new IterableOf<>(
+                () -> new Iterator<X>() {
+                    private Unchecked<I> current = new Unchecked<>(
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/Paged.java:65]: @andreoss I prefer all object's attributes declared as `final`. As this attribute needs change, you can use `AtomicReference` to achieve it.

File: src/test/java/org/cactoos/iterable/PagedTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * Test case for {@link Paged}.
+ * @since 0.47
+ * @checkstyle ClassDataAbstractionCoupling (2 lines)
+ */
+final class PagedTest {
+
+    @Test
+    @SuppressWarnings({"unchecked", "PMD.AvoidDuplicateLiterals"})
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/iterable/PagedTest.java:43]: @andreoss I think you can substitute the literals and remove `PMD.AvoidDuplicateLiterals` suppress.

File: src/test/java/org/cactoos/iterable/PagedTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import org.hamcrest.core.IsEqual;
+import org.junit.jupiter.api.Test;
+import org.llorllale.cactoos.matchers.Assertion;
+
+/**
+ * Test case for {@link Paged}.
+ * @since 0.47
+ * @checkstyle ClassDataAbstractionCoupling (2 lines)
+ */
+final class PagedTest {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/iterable/PagedTest.java:40]: @andreoss please, can you add a test when `pages.iterator().next()` throw ` NoSuchElementException`?

**Review by fabriciofx — CHANGES_REQUESTED** [review]
@andreoss I finished my review. Can you take a look, please?

File: src/test/java/org/cactoos/iterable/PagedTest.java (hunk 4 above)
Comment (andreoss) [reply src/test/java/org/cactoos/iterable/PagedTest.java:44]: @fabriciofx Done.

File: src/test/java/org/cactoos/iterable/PagedTest.java (hunk 3 above)
Comment (andreoss) [reply src/test/java/org/cactoos/iterable/PagedTest.java]: @fabriciofx Done.

File: src/main/java/org/cactoos/iterable/Paged.java (hunk 2 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Paged.java]: @victornoel Done.

**Comment by fabriciofx** [comment]
@andreoss it seems fine to me. @victornoel can you merge it, please?

File: src/main/java/org/cactoos/iterable/Paged.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+     * @param <I> Custom iterator
+     */
+    @SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
+    public <I extends Iterator<X>> Paged(
+        final Scalar<I> first, final Func<I, I> next
+    ) {
+        // @checkstyle AnonInnerLengthCheck (30 lines)
+        super(
+            new IterableOf<>(
+                () -> new Iterator<X>() {
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Paged.java:65]: @andreoss let's extract this in its own `Paged` iterator implementation and use it here

Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Paged.java:65]: @victornoel `iterator.Paged` added as a todo note, as discussed here https://github.com/yegor256/cactoos/pull/1448#discussion_r485922383

Comment (victornoel) [reply src/main/java/org/cactoos/iterable/Paged.java:65]: @andreoss haha, indeed, sorry, at least I am consistent :)

**Comment by victornoel** [comment]
@rultor merge

**Comment by victornoel** [comment]
@andreoss there are now merge conflicts sure to other PRs being merged

**Comment by andreoss** [comment]
@victornoel Fixed conflicts

**Comment by victornoel** [comment]
@rultor merge

**Pull request merged**